File: launch.R
#devtools::install_github("hrbrmstr/curlconverter")

library(curlconverter)
library(data.table)

api_proxy <- Sys.getenv("DOMINO_API_PROXY")
project_id <- Sys.getenv("DOMINO_PROJECT_ID")

hardware_tiers <- fread("/mnt/code/job_hardware_tiers.csv", header = FALSE)
for (row in 1:nrow(hardware_tiers)){
  # Call Domino's REST API via the API proxy to launch jobs
  # The API proxy is available in Domino workspaces and jobs
  # https://docs.dominodatalab.com/en/latest/api_guide/8c929e/rest-api-reference/#_startJob
  httr::POST(paste0(api_proxy, '/api/jobs/v1/jobs'),
              accept_json(),
              body = list(
                #commitId = "960a4c99a4cc38194cbacbcce41caa68ba5369ea",
                #computeCluster = list(
                  #clusterType = "dask",
                  #computeEnvironmentId = "623139857a0af0281c01a6a4",
                  #computeEnvironmentRevisionSpec = "ActiveRevision | LatestRevision | SomeRevision(623131577a0af0281c01a69a)",
                  #masterHardwareTierId = "medium-k8s",
                  #maxWorkerCount = 10,
                  #workerCount = 4,
                  #workerHardwareTier = "large-k8s",
                  #workerStorageMB = 5
                #),
                #environmentId = "623131507a0af0281c01a699",
                #environmentRevisionSpec = "ActiveRevision | LatestRevision | SomeRevision(623131577a0af0281c01a69a)",
                #externalVolumeMountIds = array(
                  #"6231327c7a0af0281c01a69b",
                  #"623132867a0af0281c01a69c"
                #),
                # Domino's REST API accepts hardware tiers' *ID* only
                hardwareTier = toString(hardware_tiers[row, 3]),
                #mainRepoGitRef = list(
                  #refType = "head | commitId | tags | branches",
                  #value = "my-test-branch"
                #),
                projectId = project_id,
                runCommand = paste("aio.R", toString(row), sep = " "),
                #snapshotDatasetsOnCompletion = false,
                title = paste("Parallel R", toString(row), sep = " ")
              ),
              encode = "json")
}

File: launch.py
import csv
import os
from domino import Domino

project = str(os.environ['DOMINO_PROJECT_OWNER'] + "/" + os.environ['DOMINO_PROJECT_NAME'])

# Tell Domino which project to start the jobs in in the format "<username>/<project_name>"
# https://docs.dominodatalab.com/en/latest/api_guide/c5ef26/the-python-domino-library/#_example
domino = Domino(project)

with open("./job_hardware_tiers.csv", 'r') as f:
    hardware_tiers_csv = csv.reader(f)
    for row in hardware_tiers_csv:
        job_number = str(row[0])
        # python-domino uses the hardware tier's display name
        hardware_tier = str(row[1]).strip()
        job_title = str("Parallel R " + job_number)
        command = str("aio.R " + job_number)
        # Use python-domino's job_start function
        # https://docs.dominodatalab.com/en/latest/api_guide/c5ef26/the-python-domino-library/#_jobs
        domino.job_start(
           command,
           hardware_tier_name=hardware_tier,
           title=job_title,
           #commit_id=None,
           #environment_id=None,
           #on_demand_spark_cluster_properties=None,
           #compute_cluster_properties=None,
           #external_volume_mounts=None
        )

